### PR TITLE
Add behavioral stats and better logging

### PR DESCRIPTION
**Pull request by portante**

### CODE REVIEW

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 1]
```diff
@@ -110,9 +110,39 @@ def parse_namespace_metadata(namespace_object)
     def get_pod_metadata(namespace_name, pod_name)
       begin
         metadata = @client.get_pod(pod_name, namespace_name)
-        return if !metadata
-        return parse_pod_metadata(metadata)
-      rescue KubeException
+        if !metadata
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: replace `if !metadata` with `unless`

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 2]
```diff
@@ -110,9 +110,39 @@ def parse_namespace_metadata(namespace_object)
...
+        @pod_cache_api_nil_error += 1
+        log.debug "Exception encountered fetching pod metadata from Kubernetes API #{@apiVersion} endpoint #{@kubernetes_url}: #{kube_error.message}"
+        nil
+      end
+    end
+
+    def get_namespace_metadata(namespace_name)
+      begin
+        metadata = @client.get_namespace(namespace_name)
+        if !metadata
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: unless

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 3]
```diff
@@ -110,9 +110,39 @@ def parse_namespace_metadata(namespace_object)
...
+            nil
+          end
+      rescue KubeException => kube_error
+        @pod_cache_api_nil_error += 1
+        log.debug "Exception encountered fetching pod metadata from Kubernetes API #{@apiVersion} endpoint #{@kubernetes_url}: #{kube_error.message}"
+        nil
+      end
+    end
+
+    def get_namespace_metadata(namespace_name)
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: code style... ruby does not typically use 'get' or 'set'

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 4]
```diff
@@ -128,6 +158,8 @@ def configure(conf)
       require 'active_support/core_ext/object/blank'
       require 'lru_redux'
 
+      log.debug "k8smd: configure() start"
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: single tick marks when there is no string substitution

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 5]
```diff
@@ -140,13 +172,35 @@ def configure(conf)
       if @cache_ttl < 0
         @cache_ttl = :none
       end
-      @cache = LruRedux::TTL::ThreadSafeCache.new(@cache_size, @cache_ttl)
+      @pod_cache = LruRedux::TTL::ThreadSafeCache.new(@cache_size, @cache_ttl)
+      @pod_cache_watch_updates = 0
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: Any reason we cant make this string of attributes into a hash? and initialize them all.  This is just a bunch of counters/metrics right?  We could create a simple 'counter' class

```
class Counter
  def initialize(name)
     @name = name
     @watch_misses = 0
     ....
  end
  attr_reader :name
  attr_reader :watch_misses

  def watch_missed
    @watch_misses += 1
  end
  ....
end

pod_cache_watch_counter = Counter.new('pod_cache')
pod_cache_watch_counter.watch_missed

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 6]
```diff
@@ -370,54 +425,97 @@ def match_annotations(annotations)
...
+        message = "start_namespace_watch: Exception encountered setting up namespace watch from Kubernetes API #{@apiVersion} endpoint #{@kubernetes_url}: #{e.message}"
+        if e.respond_to?(:response)
+          message += " (#{e.response})"
+        end
+        log.debug message
+        raise Fluent::ConfigError, message
+      end
+
       watcher.each do |notice|
-        puts notice
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: Maybe this should be replaced with an info or debug message?

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 7]
```diff
@@ -370,54 +425,97 @@ def match_annotations(annotations)
...
+            @namespace_cache_watch_ignored += 1
+        end
+      end
+    end
+
+    def dump_stats
+      @curr_time = Time.now
+      return if @curr_time - @prev_time < @stats_interval
+      @prev_time = @curr_time
+      if @use_journal
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: You can actually compact with:

```
log.info(msg) if @use_journal
```

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 8]
```diff
@@ -370,54 +425,97 @@ def match_annotations(annotations)
...
+        log.info "container name match failed: #{@container_name_match_failed}, container name id missing: #{@container_name_id_missing}"
+      end
+      if @merge_json_log
+        log.info "merge json parse errors: #{@merge_json_parse_errors}"
+      end
+      if @kubernetes_url.present?
+        log.info "pod cache stats api: updates: #{@pod_cache_api_updates}, nil_not_found: #{@pod_cache_api_nil_not_found}, nil_bad_resp_payload: #{@pod_cache_api_nil_bad_resp_payload}, nil_error: #{@pod_cache_api_nil_error}"
```
Comment (jcantrill) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: Adding a Counter class would allow you to create a `to_s` and just dump the counter

Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: https://github.com/balgarath/ruby-counter-class

File: lib/fluent/plugin/filter_kubernetes_metadata.rb (hunk 3 above)
Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: it is consistent with the `get_pod_metadata` method.

File: lib/fluent/plugin/filter_kubernetes_metadata.rb (hunk 5 above)
Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: https://github.com/balgarath/ruby-counter-class

File: lib/fluent/plugin/filter_kubernetes_metadata.rb [hunk 9]
```diff
@@ -218,10 +271,11 @@ def configure(conf)
         begin
           @annotations_regexps << Regexp.compile(regexp)
         rescue RegexpError => e
-          log.error "Error: invalid regular expression in annotation_match: #{e}"
+          log.warn "Error: invalid regular expression in annotation_match: #{e}"
```
Comment (richm) [inline lib/fluent/plugin/filter_kubernetes_metadata.rb]: Why change this from `error` to `warn`?  e.g. this is something I would want to know about asap if I typed in an incorrect regexp.

File: lib/fluent/plugin/filter_kubernetes_metadata.rb (hunk 6 above)
Comment (richm) [reply lib/fluent/plugin/filter_kubernetes_metadata.rb]: I would say `trace` level - too verbose otherwise

**Comment by portante** [comment]
Closing in favor of #89.

**Pull request closed without merging**